src/server/worker/pipeline/transform/worker.go

```diff
@@ -627,20 +627,26 @@ func fetchChunkFromWorker(driver driver.Driver, logger logs.TaggedLogger, addres
 	return grpcutil.NewStreamingBytesReader(getChunkClient, cancel), nil
 }
 
-func fetchChunk(driver driver.Driver, logger logs.TaggedLogger, info *HashtreeInfo, shard int64, stats bool) (io.ReadCloser, error) {
+func fetchChunk(driver driver.Driver, logger logs.TaggedLogger, cache *hashtree.MergeCache, chunkID string, info *HashtreeInfo, shard int64, stats bool) error {
 	if info.Address != "" {
-		reader, err := fetchChunkFromWorker(driver, logger, info.Address, info.SubtaskID, shard, stats)
+		err := func() error {
+			reader, err := fetchChunkFromWorker(driver, logger, info.Address, info.SubtaskID, shard, stats)
+			if err != nil {
+				return err
+			}
+			return cache.Put(chunkID, reader)
+		}()
 		if err == nil {
-			return reader, nil
+			return nil
 		}
 		logger.Logf("error when fetching cached chunk (%s) from worker (%s) - fetching from object store instead: %v", info.Object, info.Address, err)
 	}
 
 	reader, err := driver.PachClient().DirectObjReader(info.Object)
 	if err != nil {
-		return nil, errors.EnsureStack(err)
+		return err
 	}
-	return reader, nil
+	return cache.Put(chunkID, reader)
```

Comment:
I think you still need to close the reader in both cases here (fetch from worker and direct read).